controller/volume_clone_controller.go

```diff
@@ -0,0 +1,267 @@
+
+func (vcc *VolumeCloneController) worker() {
+	for vcc.processNextWorkItem() {
+	}
+}
+
+func (vcc *VolumeCloneController) processNextWorkItem() bool {
+	key, quit := vcc.queue.Get()
+	if quit {
+		return false
+	}
+	defer vcc.queue.Done(key)
+	err := vcc.syncHandler(key.(string))
+	vcc.handleErr(err, key)
+	return true
+}
+
+func (vcc *VolumeCloneController) handleErr(err error, key interface{}) {
+	if err == nil {
+		vcc.queue.Forget(key)
+		return
+	}
+
+	vcc.logger.WithError(err).Warnf("Error syncing Longhorn volume %v", key)
+	vcc.queue.AddRateLimited(key)
+	return
+}
+
+func (vcc *VolumeCloneController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: failed to sync volume %v", vcc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != vcc.namespace {
+		return nil
+	}
+	return vcc.reconcile(name)
+}
+
+func (vcc *VolumeCloneController) reconcile(volName string) (err error) {
+	vol, err := vcc.ds.GetVolume(volName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	if !vcc.isResponsibleFor(vol) {
+		return nil
+	}
+
+	va, err := vcc.ds.GetLHVolumeAttachment(types.GetLHVolumeAttachmentNameFromVolumeName(volName))
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		vcc.enqueueVolumeAfter(vol, constant.LonghornVolumeAttachmentNotFoundRetryPeriod)
+		return nil
+	}
+	existingVA := va.DeepCopy()
+	defer func() {
+		if err != nil {
+			return
+		}
+		if reflect.DeepEqual(existingVA.Spec, va.Spec) {
+			return
+		}
+
+		if _, err = vcc.ds.UpdateLHVolumeAttachment(va); err != nil {
+			return
+		}
+	}()
+
+	expectedAttachmentTickets := make(map[string]bool)
+
+	// case 1: this volume is target of a clone
+	if isTargetVolumeOfCloning(vol) {
+		cloningAttachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeVolumeCloneController, volName)
+		vcc.createOrUpdateAttachmentTicket(cloningAttachmentTicketID, vol.Status.OwnerID, va, longhorn.TrueValue)
+		expectedAttachmentTickets[cloningAttachmentTicketID] = true
+	}
+
+	// case 2: this volume is source of a clone
+	vols, err := vcc.ds.ListVolumes()
+	if err != nil {
+		return err
+	}
+	for _, v := range vols {
+		attachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeVolumeCloneController, v.Name)
+		if isTargetVolumeOfCloning(v) && types.GetVolumeName(v.Spec.DataSource) == vol.Name {
+			vcc.createOrUpdateAttachmentTicket(attachmentTicketID, vol.Status.OwnerID, va, longhorn.AnyValue)
+			expectedAttachmentTickets[attachmentTicketID] = true
+		}
+	}
+
+	// Delete unexpected attachment tickets
+	for attachmentTicketID, attachmentTicket := range va.Spec.AttachmentTickets {
+		if attachmentTicket.Type == longhorn.AttacherTypeVolumeCloneController {
+			if _, ok := expectedAttachmentTickets[attachmentTicketID]; !ok {
+				delete(va.Spec.AttachmentTickets, attachmentTicketID)
+			}
+		}
+	}
```

Comment:
What's the primary reason to delete other clone attachment tickets? Do you mean cloning is exclusive? What will happen if there are two volumes are being cloned from the same volume?

Reply:
I mean when the clone finished, we should delete the attachment ticket from bot source and target volume

Reply:
> What will happen if there are two volumes are being cloned from the same volume?

This should works with the current implementation of this PR

Reply:
To clarify a bit,  the two cloning will be sequential, right? It seems there is only one attachment ticket active at a time except that there is a migrating volume.

Reply:
They can be parallel because 2 clone tickets request the same node with same parameters so they can be satisfied at the same time